packages/rest/src/lib/REST.ts

```diff
@@ -0,0 +1,134 @@
+import { mergeDefault } from '@sapphire/utilities';
+import { EventEmitter } from 'events';
+import { CDN } from './CDN';
+import { InternalRequest, RequestData, RequestManager, RequestMethod } from './RequestManager';
+import { DefaultRestOptions } from './utils/constants';
+
+/**
+ * The events that the REST manager emits
+ */
+export const enum RESTEvents {
+	Debug = 'restDebug',
+	RateLimited = 'rateLimited',
+}
+
+/**
+ * Options to be passed when creating the REST instance
+ */
+export interface RESTOptions {
+	/**
+	 * The base api path, without version
+	 *
+	 * @default 'https://discord.com/api'
+	 */
+	api: string;
+	/**
+	 * The cdn path
+	 *
+	 * @default 'https://cdn.discordapp.com'
+	 */
+	cdn: string;
+	/**
+	 * The extra offset to add to ratelimits in milliseconds
+	 *
+	 * @default 50
+	 */
+	offset: number;
+	/**
+	 * The number of retries for errors with the 500 code, or errors
+	 * that timeout
+	 *
+	 * @default 3
+	 */
+	retries: number;
+	/**
+	 * The time to wait in milliseconds before a request is aborted
+	 *
+	 * @default 15_000
+	 */
+	timeout: number;
+	/**
+	 * Extra information to add to the user agent
+	 * @default `Node.js ${process.version}`
+	 */
+	userAgentAppendix: string;
+	/**
+	 * The version of the API to use
+	 *
+	 * @default '8'
+	 */
+	version: string;
+}
+
+export class REST extends EventEmitter {
+	public readonly options: RESTOptions;
+	public readonly cdn: CDN;
+	public readonly requestManager: RequestManager;
+
+	public constructor(options: Partial<RESTOptions> = {}) {
+		super();
+		this.options = mergeDefault(DefaultRestOptions, options);
+		this.options.offset = Math.max(0, this.options.offset);
```

Comment:
We shouldn't mutate options, instead, use object spread.

Reply:
Doesn't mergeDefault already cover this?

Reply:
> Doesn't mergeDefault already cover this?

[Nope](https://github.com/sapphire-project/utilities/blob/01c440a2d5cff07effe22d64dc7929a9f9efa330/packages/utilities/src/lib/mergeDefault.ts#L46-L61).

Reply:
@kyranet Unless I'm mistaken, `mergeDefault` depends on `deepClone` which [doesn't mutate](https://github.com/sapphire-project/utilities/blob/01c440a2d5cff07effe22d64dc7929a9f9efa330/packages/utilities/src/lib/deepClone.ts#L17)